[rpc] improve general error handling in process group agent and fix flaky tests

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34413 [rpc] improve general error handling in process group agent and fix flaky tests**

In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

rohan-varma committed:
[rpc] handle exceptions in ProcessGroupAgent::enqueueRecv

In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Previously, these tests were flaky and disabled. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -401,105 +447,119 @@ void ProcessGroupAgent::enqueueSend(SendWork work) {
-          if (futureResponse->completed()) {
-            if (!futureResponse->hasError()) {
-              send(work.from_, std::move(*futureResponse).moveValue());
+void ProcessGroupAgent::handleRecv(RecvWork& work) {
```
for reviewers: there are no real changes in this logic, the code is just moved to a helper function.

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Previously, these tests were flaky and disabled. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Previously, these tests were flaky and disabled. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Previously, these tests were flaky and disabled. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -388,8 +410,32 @@ void ProcessGroupAgent::handleSend(const SendWork& work) {
+  // preemptively remove sends from other threads.
+  auto& vec = currentPendingSends_[dst];
+  for (auto& p : pendingSends) {
+    auto it = std::find(vec.begin(), vec.end(), p);
```
This is linear complexity? Will it be more efficient if we use the thread id to key pending sends in `currentPendingSends_`? Or is there any reason that they have to be keyed using `dst`?

rohan-varma replied:
I think we can key by the thread_id and then remove all sends corresponding to the thread_id here instead. Alternatively, it looks like we could also turn the vector into a set which would give us constant lookup since `pendingSends` here is only always of size 2.

mrshenli commented on torch/testing/_internal/dist_utils.py:
```diff
@@ -87,28 +88,42 @@ def new_test_method(self, *arg, **kwargs):
     indicate that the node has failed in unit tests.
+    Args:
+    rank (int) - Rank of the node expected to fail
+    expected_error_regex (optional, str): Regex of exception message expected. Useful to ensure a specific failure
```
one arg used "-" another used ":" to separate name and description.

mrshenli commented on torch/testing/_internal/dist_utils.py:
```diff
@@ -87,28 +88,42 @@ def new_test_method(self, *arg, **kwargs):
-            break
+            time.sleep(0.1)
+        except Exception as e:
+            if re.match(pattern=expected_error_regex, string=str(e)):
```
Was the intention to enforce the error message matches the given regex?

rohan-varma replied:
Yes, this is to ensure that we get a desired error message when waiting for the node failure. My use case for this is that when testing failure edge cases, we are able to push out the RPC before the other node shuts down, and then get a timed out message instead of the expected error message due to the node being shutdown. In that case we will automatically retry here.

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -3,14 +3,15 @@
 from enum import Enum
-
 import torch
+from datetime import timedelta
```
Shall we sort imports?

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -1425,22 +1417,28 @@ def test_backward_node_failure_python_udf(self):
+                                args=(),
+                            )
+                        except Exception as e:
+                            pass
```
do we have a specific error message that we can test?

rohan-varma replied:
Yes, this should be a RPC timed out error message since we never get a response back from the remote end, since it has shutdown.

rohan-varma replied:
Will update the diff to add this.

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -257,6 +257,16 @@ void ProcessGroupAgent::shutdown() {
+    for (auto& it : currentPendingSends_) {
+      const auto& pendingSends = it.second;
+      for (auto send : pendingSends) {
+        send->abort();
```
Do we need to check complete before abort? I assume aborting completed sends is OK?

rohan-varma replied:
We should be able to `abort` completed sends, though we could save this call which requires a lock and a CV notify in gloo if we check for completed.

mrshenli reviewed (approved):
Thanks for fixing! Looks good to me. I left some minor comments. Two things worth some attention:

1. lint failure is real
2. it will be great if we can avoid the linear complexity on deleting pending sends.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -257,6 +257,16 @@ void ProcessGroupAgent::shutdown() {
+    for (auto& it : currentPendingSends_) {
+      const auto& pendingSends = it.second;
+      for (auto send : pendingSends) {
+        send->abort();
```
I'll add a log giving information (sender, receiver) of the send that is being aborted. We should not reach here in a graceful shutdown, since all sends should have been satisfied in a graceful shutdown.

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

jjlilley commented:
fwiw, this LGTM as well.

There might be some minor merge conflicts with the activeCalls_ change I landed yesterday (sorry)

It would be great to get this in soon, e.g. Wanchao is running into some exception-related errors in process_group_agent with his autograd change, and suspect that this could help.

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -405,117 +456,136 @@ void ProcessGroupAgent::enqueueSend(SendWork work) {
+        try {
+          auto shouldIncr = handleRecv(work);
+          if (shouldIncr) {
+            recvCounts_.increment(work.from_.id_);
```
@mrshenli Could you take a look at this change to use of `recvCounts`? Essentially what I wanted to do is increment recvCounts if `handleRecv` returns successfully and processes the receive, or if there is an exception. We don't want to increment if the recv was dropped, due to e.g. timeout.

mrshenli replied:
The logic sounds correct to me :)

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma commented:
lint errors are in untouched file.

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] improve general error handling in process group agent and fix flaky tests"


In this diff we have made various improvements to ProcessGroupAgent in order to accomodate edge and error cases such as a "non-clean" shutdown (shutdowns in which we abort RPC as quickly as possible, and don't wait for all pending work across all RPC agents to be completed):

1. Catch and log exceptions in `enqueueRecv`. This prevents us from calling `std::terminate()` in a different thread and logs an error message indicating the issue. With this we no longer have crashes caused by exceptions in this thread during non-graceful shutdown.

2. Provide cleaner error messages everywhere (and use `c10::str` where possible). One example is in `agent::send()`.

3. Add the ability to abort pending sends that cause blocking waits in `handleSend`. The reason we need to abort this is since during a non-graceful shutdown, we could become blocked waiting for these since there is no guarantee the remote end is still active and this would result in a long wait and eventual timeout. We abort these by adding them to a map, and go through this map during `shutdown()`.

4. Fix flaky tests: `test_handle_send_exceptions` and `test_backward_node_failure` and `test_backward_node_failure_python_udf`. These tests were flaky since they dealt with non-graceful shutdown of workers which has chances for a bunch of edge cases explained above.

We have also refactored `createExceptionResponse`, `enqueueRecv`, and some test functions for the above reasons in this diff.

For testing:
Ensured that the tests are no longer flaky with 500 tests runs. Also added a unit test in the internal `ProcessGroupAgentTest.cpp`.

Differential Revision: [D20269074](https://our.internmc.facebook.com/intern/diff/D20269074/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20269074/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -428,105 +484,124 @@ void ProcessGroupAgent::enqueueSend(SendWork work) {
+  threadPool_.run(std::bind(
+      [&](RecvWork& work) {
+        try {
+          auto shouldIncr = handleRecv(work);
```
nit: (fix can come in the next PR is necessary) Looks like `shouldIncr` is only used here. Does the following work?

```c++
if (handleRecv(work)) {
    recvCounts_.increment(work.from_.id_);
}
```

rohan-varma replied:
Yes, makes sense. It only needs to be a boolean.

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -428,105 +484,124 @@ void ProcessGroupAgent::enqueueSend(SendWork work) {
-            futureResponse->setError(e.what());
-          }
-          if (futureResponse->completed()) {
+int ProcessGroupAgent::handleRecv(RecvWork& work) {
```
nit: (fix can come in the next PR is necessary) any reason for returning 0/1 instead of false/true here? Do we plan to return other values in the future?

rohan-varma replied:
This is true, it should return a bool. I will update it in the next diff.

jjlilley commented on torch/csrc/distributed/autograd/context/container.cpp:
```diff
@@ -160,11 +160,21 @@ void DistAutogradContainer::sendReleaseContextRpc(int64_t context_id) {
+  // ungraceful shutdown, where we are shutting down RPC and also processing
+  // this message in a separate thread concurrently. In this case, don't throw
+  // here.
+  try {
```
Would it be better to move this try/catch inside the for loop? that way, if one of them fails, we'll at least attempt the others.

rohan-varma replied:
@jjlilley I would like to do that, though the only reason I've seen this failing and what this was supposed to fix is when this is called while RPC is shutdown, in which we have the following 2 errors:
1) The call to `getCurrentAgent()` fails, which is outside the loop
2) `agent.send()` fails since RPC has been shutdown

So even if (2) fails inside the loop, we're certain that the rest will also fail since RPC is shutdown. 

We can probably have separate try/catches for getting the current RPC agent and then try/catches for the individual sends, I can do this in a follow up diff.

jjlilley replied:
Sure, go ahead and land this then when the tests clear, so that we can minimize merge conflicts.

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -400,8 +428,31 @@ void ProcessGroupAgent::handleSend(const SendWork& work) {
   }
+  // Write pendingSends to a global map so that they can be interrupted by
+  // ::shutdown().
+  std::unique_lock<std::mutex> pendingSendGuard(pendingSendMutex_);
```
this is fine, though might be slightly easier to follow the locking if using blocks:

```
   {
      std::unique_lock<std::mutex> pendingSendGuard(pendingSendMutex_);
      for (auto& p : pendingSends) {
        currentPendingSends_[dst].insert(p);
      }
    }
```
